src/UCTSearch.cpp

```diff
@@ -39,10 +39,29 @@
 
 using namespace Utils;
 
-UCTSearch::UCTSearch(GameState & g)
-    : m_rootstate(g) {
+UCTSearch::UCTSearch() {
     set_playout_limit(cfg_max_playouts);
     set_visit_limit(cfg_max_visits);
+    m_root = std::make_unique<UCTNode>(FastBoard::PASS, 0.0f, 0.5f);
+}
+
+void UCTSearch::set_gamestate(GameState & g) {
+    // Definition of m_playouts is playouts from a certain GameState.
+    // So reset this count now.
+    m_playouts = 0;
+    if (m_rootstate.get_komi() != g.get_komi()) {
+        m_root = std::make_unique<UCTNode>(FastBoard::PASS, 0.0f, 0.5f);
+        m_rootstate = g;
```

Comment:
Is there ever a case where you wouldn't want to do m_rootstate = g in this function? I think it can go above the ifs.

Maybe if get_hash() != get_hash()? Then I'd pull that case out to remove the code duplication.

Reply:
I refactored the code and added a comment to be clear. It avoids doing the `m_rootstate=g` copy when there is a match at the root, which happens every move in self play games.

BTW I think komi should be included in m_hash, that would get rid of this special case and also the one in NNCache. Save that change for later though. For now it doesn't matter since we only support komi=7.5.

Reply:
After a couple iterations I have a much cleaner way of doing this.